crates/miden-objects/src/account/component/template/storage/entry_content.rs

```diff
@@ -537,66 +583,98 @@ impl MapRepresentation {
+                let entries = entries
+                    .iter()
+                    .map(|map_entry| {
+                        let key = map_entry
+                            .key()
+                            .try_build_word(init_storage_data, identifier.name.clone())?;
+                        let value = map_entry
+                            .value()
+                            .try_build_word(init_storage_data, identifier.name.clone())?;
+                        Ok((key, value))
+                    })
+                    .collect::<Result<Vec<(Word, Word)>, _>>()?;
+
+                StorageMap::with_entries(entries).map_err(|err| {
+                    AccountComponentTemplateError::StorageMapHasDuplicateKeys(Box::new(err))
+                })
+            },
+            MapRepresentation::Template { identifier } => {
+                if let Some(entries) = init_storage_data.map_entries(&identifier.name) {
+                    return StorageMap::with_entries(entries.clone()).map_err(|err| {
+                        AccountComponentTemplateError::StorageMapHasDuplicateKeys(Box::new(err))
+                    });
+                }
 
-        Ok(())
+                Err(AccountComponentTemplateError::PlaceholderValueNotProvided(
+                    identifier.name.clone(),
+                ))
+            },
+        }
     }
-}
 
-impl From<MapRepresentation> for Vec<MapEntry> {
-    fn from(value: MapRepresentation) -> Self {
-        value.entries
+    /// Validates the map representation by checking for duplicate keys and placeholder validity.
+    pub(crate) fn validate(&self) -> Result<(), AccountComponentTemplateError> {
+        match self {
+            MapRepresentation::Template { .. } => Ok(()),
+            MapRepresentation::Value { entries, .. } => {
+                let mut seen_keys = BTreeSet::new();
+                for entry in entries.iter() {
+                    entry.key().validate()?;
+                    entry.value().validate()?;
+                    if let Ok(key) = entry
+                        .key()
+                        .try_build_word(&InitStorageData::default(), StorageValueName::empty())
+                        && !seen_keys.insert(key)
+                    {
+                        return Err(AccountComponentTemplateError::StorageMapHasDuplicateKeys(
+                            Box::from(format!("key `{key}` is duplicated")),
+                        ));
+                    }
+                }
+
+                Ok(())
+            },
+        }
     }
 }
 
 impl Serializable for MapRepresentation {
     fn write_into<W: ByteWriter>(&self, target: &mut W) {
-        self.entries.write_into(target);
-        target.write(&self.identifier);
+        match self {
+            MapRepresentation::Value { identifier, entries } => {
+                target.write_u8(0u8);
+                target.write(entries);
+                target.write(identifier);
```

Comment:
Nit: I would've expected identifier first, entries second because that's more like "key-value" order.